ml-agents-envs/mlagents_envs/base_env.py

```diff
@@ -247,90 +246,58 @@ def empty(spec: "BehaviorSpec") -> "TerminalSteps":
-     - observation_shapes is a List of Tuples of int : Each Tuple corresponds
-     to an observation's dimensions. The shape tuples have the same ordering as
-     the ordering of the DecisionSteps and TerminalSteps.
-     - action_type is the type of data of the action. it can be discrete or
-     continuous. If discrete, the action tensors are expected to be int32. If
-     continuous, the actions are expected to be float32.
-     - action_shape is:
-       - An int in continuous action space corresponding to the number of
-     floats that constitute the action.
-       - A Tuple of int in discrete action space where each int corresponds to
-       the number of discrete actions available to the agent.
-    """
-
-    observation_shapes: List[Tuple]
-    action_type: ActionType
-    action_shape: Union[int, Tuple[int, ...]]
+class ActionSpec(NamedTuple):
+    num_continuous_actions: int
+    discrete_branch_sizes: Tuple[int, ...]
 
+    # For backwards compatibility
     def is_action_discrete(self) -> bool:
         """
         Returns true if this Behavior uses discrete actions
         """
-        return self.action_type == ActionType.DISCRETE
+        return self.discrete_action_size > 0
 
+    # For backwards compatibility
     def is_action_continuous(self) -> bool:
         """
         Returns true if this Behavior uses continuous actions
         """
-        return self.action_type == ActionType.CONTINUOUS
+        return self.continuous_action_size > 0
+
+    @property
+    def discrete_action_branches(self) -> Tuple[int, ...]:
+        return self.discrete_branch_sizes  # type: ignore
+
+    @property
+    def discrete_action_size(self) -> int:
+        return len(self.discrete_branch_sizes)
+
+    @property
+    def continuous_action_size(self) -> int:
+        return self.num_continuous_actions
 
     @property
     def action_size(self) -> int:
-        """
-        Returns the dimension of the action.
-         - In the continuous case, will return the number of continuous actions.
-         - In the (multi-)discrete case, will return the number of action.
-         branches.
-        """
-        if self.action_type == ActionType.DISCRETE:
-            return len(self.action_shape)  # type: ignore
-        else:
-            return self.action_shape  # type: ignore
+        return self.discrete_action_size + self.continuous_action_size
 
     @property
-    def discrete_action_branches(self) -> Optional[Tuple[int, ...]]:
-        """
-        Returns a Tuple of int corresponding to the number of possible actions
-        for each branch (only for discrete actions). Will return None in
-        for continuous actions.
-        """
-        if self.action_type == ActionType.DISCRETE:
-            return self.action_shape  # type: ignore
-        else:
-            return None
+    def total_action_size(self) -> int:
+        return sum(self.discrete_action_branches) + self.continuous_action_size
```

Comment:
```suggestion
        return self.num_continuous_actions + self.continuous_action_size
```

Reply:
Might need comments on why both this and action_size exist.

Reply:
I don't understand the suggested change. Wouldn't that be 2*continuous_action_size. I'll add a comment

Reply:
Sorry, mistake in the proposed change, I meant `discrete_action_size` but now I see they're different.